internal/uvm/virtual_device.go

```diff
@@ -10,57 +11,110 @@ import (
+// and the VPCIDevice is returned.
+// Otherwise, a new request is made to assign the target device indicated by the deviceID
+// onto the UVM. A new VPCIDevice entry is made on the UVM and the VPCIDevice is returned
+// to the caller
+func (uvm *UtilityVM) AssignDevice(ctx context.Context, deviceID string) (*VPCIDevice, error) {
+	if uvm.operatingSystem == "windows" {
+		return nil, errors.New("assigned devices is not currently supported on wcow")
+	}
+
 	guid, err := guid.NewV4()
 	if err != nil {
 		return nil, err
 	}
-	id := guid.String()
+	vmBusGUID := guid.String()
 
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	if err := uvm.modify(ctx, &hcsschema.ModifySettingRequest{
-		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, id),
+
+	existingVPCIDevice := uvm.vpciDevices[deviceID]
+	if existingVPCIDevice != nil {
+		existingVPCIDevice.refCount++
+		return existingVPCIDevice, nil
+	}
+
+	targetDevice := hcsschema.VirtualPciDevice{
+		Functions: []hcsschema.VirtualPciFunction{
+			{
+				DeviceInstancePath: deviceID,
+			},
+		},
+	}
+
+	request := &hcsschema.ModifySettingRequest{
+		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, vmBusGUID),
 		RequestType:  requesttype.Add,
-		Settings:     device,
-		GuestRequest: guestrequest.GuestRequest{
+		Settings:     targetDevice}
+
+	if uvm.operatingSystem != "windows" {
+		// for LCOW, we need to make sure that specific paths relating to the
+		// device exist so they are ready to be used by later
+		// work in openGCS
+		request.GuestRequest = guestrequest.GuestRequest{
 			ResourceType: guestrequest.ResourceTypeVPCIDevice,
 			RequestType:  requesttype.Add,
 			Settings: guestrequest.LCOWMappedVPCIDevice{
-				VMBusGUID: id,
+				VMBusGUID: vmBusGUID,
 			},
-		},
-	}); err != nil {
+		}
+	}
+
+	if err := uvm.modify(ctx, request); err != nil {
 		return nil, err
 	}
-	return &VPCIDevice{
-		vm: uvm,
-		ID: id,
-	}, nil
+	result := &VPCIDevice{
+		vm:               uvm,
+		VMBusGUID:        vmBusGUID,
+		deviceInstanceID: deviceID,
+		refCount:         1,
+	}
+	uvm.vpciDevices[deviceID] = result
+	return result, nil
 }
 
-// RemoveDevice removes a vpci device from the uvm
-func (uvm *UtilityVM) RemoveDevice(ctx context.Context, id string) error {
+// removeDevice removes a vpci device from a uvm when there are
+// no more references to a given VPCIDevice. Otherwise, decrements
+// the reference count of the stored VPCIDevice and returns nil.
+func (uvm *UtilityVM) removeDevice(ctx context.Context, deviceInstanceID string) error {
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	return uvm.modify(ctx, &hcsschema.ModifySettingRequest{
-		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, id),
-		RequestType:  requesttype.Remove,
-	})
+
+	vpci := uvm.vpciDevices[deviceInstanceID]
+	if vpci == nil {
+		return fmt.Errorf("no device with ID %s is present on the uvm %s", deviceInstanceID, uvm.ID())
+	}
+
+	vpci.refCount--
+	if vpci.refCount == 0 {
+		uvm.vpciDevices[deviceInstanceID] = nil
```

Comment:
I think it's better to do `delete(uvm.vpciDevices, deviceInstanceID)` here. Otherwise the map might keep a record of the key.

Reply:
I imagine we have other cases that should be updated to this as well.

Reply:
for other resources?

Reply:
yes

Reply:
I see. I've updated it here at least :)